src/Temporalio/Bridge/CustomSlotSupplier.cs

```diff
@@ -0,0 +1,246 @@
+        private uint permitId = 1;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CustomSlotSupplier" /> class.
+        /// </summary>
+        /// <param name="userSupplier">User's slot supplier implementation'.</param>
+        /// <param name="loggerFactory">Logger factory.</param>
+        internal unsafe CustomSlotSupplier(
+            Temporalio.Worker.Tuning.CustomSlotSupplier userSupplier,
+            ILoggerFactory loggerFactory)
+        {
+            this.logger = loggerFactory.CreateLogger<CustomSlotSupplier>();
+            this.userSupplier = userSupplier;
+
+            var interopCallbacks = new Interop.CustomSlotSupplierCallbacks
+            {
+                reserve = FunctionPointer<Interop.CustomReserveSlotCallback>(Reserve),
+                cancel_reserve = FunctionPointer<Interop.CustomCancelReserveCallback>(CancelReserve),
+                try_reserve = FunctionPointer<Interop.CustomTryReserveSlotCallback>(TryReserve),
+                mark_used = FunctionPointer<Interop.CustomMarkSlotUsedCallback>(MarkUsed),
+                release = FunctionPointer<Interop.CustomReleaseSlotCallback>(Release),
+                free = FunctionPointer<Interop.CustomSlotImplFreeCallback>(Free),
+            };
+
+            PinCallbackHolder(interopCallbacks);
+        }
+
+        private static Temporalio.Worker.Tuning.SlotInfo SlotInfoFromBridge(Interop.SlotInfo slotInfo)
+        {
+            return slotInfo.tag switch
+            {
+                Interop.SlotInfo_Tag.WorkflowSlotInfo =>
+                    new Temporalio.Worker.Tuning.SlotInfo.WorkflowSlotInfo(
+                        ByteArrayRef.ToUtf8(slotInfo.workflow_slot_info.workflow_type), slotInfo.workflow_slot_info.is_sticky != 0),
+                Interop.SlotInfo_Tag.ActivitySlotInfo =>
+                    new Temporalio.Worker.Tuning.SlotInfo.ActivitySlotInfo(
+                        ByteArrayRef.ToUtf8(slotInfo.activity_slot_info.activity_type)),
+                Interop.SlotInfo_Tag.LocalActivitySlotInfo =>
+                    new Temporalio.Worker.Tuning.SlotInfo.LocalActivitySlotInfo(
+                        ByteArrayRef.ToUtf8(slotInfo.local_activity_slot_info.activity_type)),
+                _ => throw new System.ArgumentOutOfRangeException(nameof(slotInfo)),
+            };
+        }
+
+        private unsafe void Reserve(Interop.SlotReserveCtx* ctx, void* sender)
+        {
+            SafeReserve(new IntPtr(ctx), new IntPtr(sender));
+        }
+
+        // Note that this is always called by Rust, either because the call is cancelled or because
+        // it completed. Therefore the GCHandle is always freed.
+        private unsafe void CancelReserve(void* tokenSrc)
+        {
+            var handle = GCHandle.FromIntPtr(new IntPtr(tokenSrc));
+            var cancelTokenSrc = (CancellationTokenSource)handle.Target!;
```

Comment:
To confirm, this can never be called before `set_reserve_cancel_target` returns and never called after `complete_async_reserve` returns right?